spec/notification_center_spec.rb

```diff
@@ -0,0 +1,203 @@
+        expect(notification_center.add_notification_listener(
+                 Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+                 @callback_reference
+        )).to eq(1)
+      end
+
+      it 'shoud return -1 when callback already exists' do
+        notification_center = Optimizely::NotificationCenter.new(Optimizely::SimpleLogger.new)
+        notification_center.add_notification_listener(
+          Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+          @callback_reference
+        )
+        expect(notification_center.add_notification_listener(
+                 Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+                 @callback_reference
+        )).to eq(-1)
+      end
+    end
+
+    describe '.remove_notification_listener' do
+      before(:example) do
+        @notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION]
+        @notification_center = Optimizely::NotificationCenter.new(spy_logger)
+        @notification_center.add_notification_listener(@notification_type, @callback_reference)
+      end
+
+      it 'shoud log and return nil when notification id is blank' do
+        expect(@notification_center.remove_notification_listener(nil)).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+          .with(Logger::ERROR, 'Notification id can not be blank!')
+      end
+
+      it 'shoud return true if notification is removed' do
+        expect(@notification_center.remove_notification_listener(1)).to eq(true)
+      end
+
+      it 'shoud return false if notification failed to remove' do
+        expect(@notification_center.remove_notification_listener(2))
+          .to eq(false)
+      end
+    end
+
+    describe '.clear_notifications' do
+      it 'should return log of notifications cleared' do
+        notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION]
+        notification_center = Optimizely::NotificationCenter.new(spy_logger)
+        notification_center.add_notification_listener(notification_type, @callback_reference)
+        notification_center.clear_notifications(notification_type)
+        expect(spy_logger).to have_received(:log).once
+          .with(Logger::INFO, "All callbacks for notification type #{notification_type} have been removed.")
+      end
+    end
+
+    describe '.fire_notifications' do
+      before(:example) do
+        config = Optimizely::ProjectConfig.new(@config_body_json, @logger, @error_handler)
+        @event_builder = Optimizely::EventBuilder.new(config)
+        @args = [
+          config.get_experiment_from_key('test_experiment'),
+          'test_user',
+          {},
+          '111128',
+          @event_builder.create_impression_event(
+            config.get_experiment_from_key('test_experiment'),
+            '111128', 'test_user', nil
+          )
+        ]
+      end
+
+      it 'should return success log for notification sent' do
```

Comment:
See PHP-SDK for fireNotification methods.
Mock actual method being called rather than asserting log message

Reply:
Also write a test case where
2 callbacks added for Type A,
1 callback added for TypeB
fireNotification for Type A and verify that
 - both callbacks for Type A are called
 - callback for Type B isn't called

Reply:
Done

Reply:
Done